bindings/rust/src/common.rs

```diff
@@ -0,0 +1,76 @@
+use std::ffi::CString;
+use crate::c_bindings;
+
+#[allow(dead_code)]
+pub struct DeviceDefinition {
+    pub def: c_bindings::InputtinoDeviceDefinition,
+    // Keep those around since we are passing them as pointers
+    name: CString,
+    phys: CString,
+    uniq: CString,
+}
+
+impl DeviceDefinition {
+    pub fn new(name: &str, vendor_id: u16, product_id: u16, version: u16, phys: &str, uniq: &str) -> Self {
+        let name = CString::new(name).unwrap();
+        let phys = CString::new(phys).unwrap();
+        let uniq = CString::new(uniq).unwrap();
+        let def = c_bindings::InputtinoDeviceDefinition {
+            name: name.as_ptr(),
+            vendor_id: vendor_id,
+            product_id: product_id,
+            version: version,
+            device_phys: phys.as_ptr(), // TODO: optional, if not present random MAC address
+            device_uniq: uniq.as_ptr(),
+        };
+        DeviceDefinition { def, name, phys, uniq }
+    }
+}
+
+pub unsafe extern "C" fn error_handler_fn(error_message: *const ::core::ffi::c_char,
+                                          user_data: *mut ::core::ffi::c_void) {
+    let error_str = std::ffi::CStr::from_ptr(error_message);
+    let user_data = user_data as *mut CString;
+    *user_data = CString::from(error_str);
+}
+
+
+#[macro_export]
+macro_rules! get_nodes {
+    ( $fn_call:expr,$var:expr ) => {
+        {
+            let mut nodes_count: core::ffi::c_int = 0;
+            let nodes = $fn_call($var, &mut nodes_count);
+            if nodes.is_null() {
+                return Err("Failed to get nodes".to_string());
+            }
+
+            let mut result = Vec::new();
+            for i in 0..nodes_count {
+                let node = std::ffi::CString::from_raw(*nodes.offset(i as isize));
+                result.push(node.to_str().unwrap().to_string());
+            }
+            Ok(result)
+        }
+    };
+}
+
+#[macro_export]
```

Comment:
Sorry if I'm missing something, but why is this macro exported? It seems like it's just an internal wrapper. I also suspect it doesn't need to be a macro, since the `$fn_call` always has the same (generic) signature.

Reply:
:shrug: good question. I didn't touch this part (most of this comes from the `dev-rust` branch from @ABeltramo, which this PR builds upon).

Reply:
Without seeing the signature for the `make_foo` fns, I think this can just be 

```rust
fn make_device<T, F>(def: &DeviceDefinition, f: F) -> T
where F: FnOnce(thingy) -> T
```

Reply:
I think 🤷 perfectly describe it for me too. No specific reason probably, just me messing around with Rust, feel free to refactor.

Reply:
I replaced both macros with a function.